tests/functional/tests/surprise_shutdown/test_management_surprise_shutdown.py

```diff
@@ -0,0 +1,682 @@
+            assert stats["usage"]["occupancy"]["value"] == 1
+            core = Core(device=core_device, try_add=True)
+            cache.add_core(core)
+            assert ocf_read(cache, core, io_offset) == 0xAA
+
+        cache.stop()
+
+        # advance error injection point
+        error_io_seq_no += 1
+
+
+@pytest.mark.security
+def test_surprise_shutdown_cache_reinit(pyocf_ctx):
+    core_device = Volume(S.from_MiB(10))
+
+    error_io = {IoDir.WRITE: 0}
+
+    io_offset = mngmt_op_surprise_shutdown_test_io_offset
+
+    error_triggered = True
+    while error_triggered:
+        # Start cache device without error injection
+        device = ErrorDevice(
+            mngmt_op_surprise_shutdown_test_cache_size, error_seq_no=error_io, armed=False
+        )
+
+        # start WB
+        cache = Cache.start_on_device(device, cache_mode=CacheMode.WB)
+        core = Core(device=core_device, try_add=False)
+        cache.add_core(core)
+
+        # insert dirty cacheline
+        ocf_write(cache, core, 0xAA, io_offset)
+
+        cache.stop()
+
+        assert core_device.get_bytes()[io_offset] == Volume.VOLUME_POISON
+
+        # start error injection
+        device.arm()
+
+        # power failure during cache re-initialization
+        try:
+            # sets force = True by default
+            cache = Cache.start_on_device(device, cache_mode=CacheMode.WB)
+            status = OcfErrorCode.OCF_OK
+        except OcfError as ex:
+            status = ex.error_code
+            cache = None
+
+        error_triggered = device.error_triggered()
+        assert error_triggered == (status == OcfErrorCode.OCF_ERR_WRITE_CACHE)
+
+        if cache:
+            with pytest.raises(OcfError) as ex:
+                cache.stop()
+            assert ex.value.error_code == OcfErrorCode.OCF_ERR_WRITE_CACHE
+
+        device.disarm()
+
+        cache = Cache.load_from_device(device)
+
+        stats = cache.get_stats()
+        if stats["conf"]["core_count"] == 0:
+            cache.add_core(core)
+            assert ocf_read(cache, core, io_offset) == Volume.VOLUME_POISON
+
+        cache.stop()
+
+        error_io[IoDir.WRITE] += 1
+
+
+def _test_surprise_shutdown_mngmt_generic(pyocf_ctx, func):
+    core_device = Volume(S.from_MiB(10))
+    core = Core(device=core_device, try_add=False)
+
+    def prepare(cache):
+        cache.add_core(core)
+
+    def test(cache):
+        func(cache, core)
+        cache.save()
+
+    mngmt_op_surprise_shutdown_test(pyocf_ctx, test, prepare, None)
+
+
+@pytest.mark.security
+def test_surprise_shutdown_change_cache_mode(pyocf_ctx):
+    _test_surprise_shutdown_mngmt_generic(
+        pyocf_ctx, lambda cache, core: cache.change_cache_mode(CacheMode.WT)
+    )
+
+
+@pytest.mark.security
+def test_surprise_shutdown_set_cleaning_policy(pyocf_ctx):
+    core_device = Volume(S.from_MiB(10))
+    core = Core(device=core_device, try_add=False)
+
+    for c1 in CleaningPolicy:
+        for c2 in CleaningPolicy:
```

Comment:
Nested for loop can be replaced with `for c1, c2 in zip(CleaningPolicy, CleaningPolicy)`

Reply:
I think zip is not sufficient - It would generate only pairs of first with first, second with second and so on, you'd need `product()` from `itertools` to do that right